### PR TITLE
core/types: Populate Jovian receipt fields

### LINKED ISSUE #708: jovian: Add & populate receipt fields (daFootprintGasScalar, blobGasUsed)
Following our pattern for previous forks, we want to add a new fields to the API receipt, `daFootprintGasScalar uint16`, and populate it when requesting a receipt via the API, at
https://github.com/ethereum-optimism/op-geth/blob/96738d21616cd7c37101fd49234e361f70e7b64c/core/types/receipt.go?plain=1#L589
when Jovian is active, and also set the receipt's `blobGasUsed` field to the transaction's DA footprint.

Note that this is not a consensus change, as the receipt hash only contains these fields:
https://github.com/ethereum-optimism/op-geth/blob/96738d21616cd7c37101fd49234e361f70e7b64c/core/types/receipt.go?plain=1#L126-L132

### PR DESCRIPTION
**Pull request by sebastianst**
**Description**

Populates Jovian receipt fields during receipt field derivation: `daFootprintGasScalar` and `blobGasUsed` (to individual transaction DA footprint).

Note that most of the diff comes from moving code around into new files to reduce the upstream diff and isolate OP Stack changes more into its own files to reduce merge conflicts.
- OP Stack diff for receipts got moved from file `receipt.go` into `receipt_opstack.go`.
- For tests, blobal vars that we introduced, as well as new tests, are moved from `receipt_test.go` to `receipt_opstack_test.go`.
- Test fixture functions `getOptimism(Ecotone|Isthmus|Jovian)TxReceipts` got improved by using previous fork's function instead of being an almost identical copy of the previous function.

**Tests**

Added new test for Jovian receipts derivation. Also refactored the existing tests to be less repetitive and fixed a few things, like proper clearing of all OP Stack additional receipt fields.

**Metadata**

Fixes #708

### CODE REVIEW

File: core/types/receipt.go [hunk 1]
```diff
@@ -612,26 +614,8 @@ func (rs Receipts) DeriveFields(config *params.ChainConfig, blockHash common.Has
...
-			rs[i].L1Fee, rs[i].L1GasUsed = gasParams.costFunc(txs[i].RollupCostData())
-			rs[i].FeeScalar = gasParams.feeScalar
-			rs[i].L1BaseFeeScalar = u32ptrTou64ptr(gasParams.l1BaseFeeScalar)
-			rs[i].L1BlobBaseFeeScalar = u32ptrTou64ptr(gasParams.l1BlobBaseFeeScalar)
-			if gasParams.operatorFeeScalar != nil && gasParams.operatorFeeConstant != nil && (*gasParams.operatorFeeScalar != 0 || *gasParams.operatorFeeConstant != 0) {
-				rs[i].OperatorFeeScalar = u32ptrTou64ptr(gasParams.operatorFeeScalar)
-				rs[i].OperatorFeeConstant = gasParams.operatorFeeConstant
-			}
-		}
+	if config.IsOptimismBedrock(new(big.Int).SetUint64(blockNumber)) && len(txs) >= 2 {
```
Comment (joshklop) [inline core/types/receipt.go:617]: Condition changed: we still need the `config.Optimism != nil`, correct? Suggesting we leave the condition unchanged since it doesn't seem to have an impact on this PR. 
```suggestion
	if config.Optimism != nil && len(txs) >= 2 && config.IsOptimismBedrock(new(big.Int).SetUint64(blockNumber)) {
```

File: core/types/receipt_opstack.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		rcd := txs[i].RollupCostData()
+		rs[i].L1Fee, rs[i].L1GasUsed = gasParams.costFunc(rcd)
+		rs[i].FeeScalar = gasParams.feeScalar
+		rs[i].L1BaseFeeScalar = u32ptrTou64ptr(gasParams.l1BaseFeeScalar)
+		rs[i].L1BlobBaseFeeScalar = u32ptrTou64ptr(gasParams.l1BlobBaseFeeScalar)
+		if gasParams.operatorFeeScalar != nil && gasParams.operatorFeeConstant != nil && (*gasParams.operatorFeeScalar != 0 || *gasParams.operatorFeeConstant != 0) {
+			rs[i].OperatorFeeScalar = u32ptrTou64ptr(gasParams.operatorFeeScalar)
+			rs[i].OperatorFeeConstant = gasParams.operatorFeeConstant
+		}
+		if config.IsJovian(blockTime) {
```
Comment (joshklop) [inline core/types/receipt_opstack.go]: nit: we can have `isJovian := config.IsJovian(blockTime)` at the top of the function to avoid checking on every receipt.

File: core/types/receipt_opstack.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+			rs[i].DAFootprintGasScalar = &daFootprintGasScalar
```
Comment (joshklop) [inline core/types/receipt_opstack.go:49]: It feels a tad wasteful to include the scalar (a block-level attribute) in every receipt when it can be statically inferred from the `BlobGasUsed` and the transaction.

I see we include other block-level attributes like the `operatorFeeScalar` in receipts, although I'm not sure if they can be inferred like the `daFootprintGasScalar`.

Any thoughts?

File: core/types/receipt_opstack.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// deriveOPStackFields derives the OP Stack specific fields for each receipt.
+// It must only be called for blocks with at least two transactions.
```
Comment (joshklop) [inline core/types/receipt_opstack.go]: nit: it will still work even if there is only the system deposit transaction right? 
```suggestion
// It must only be called for blocks with at least one transaction (the system deposit).
```

**Comment by joshklop** [comment]
This is a bit challenging to review since there are intersecting refactors in the same commit as the consensus changes. If you don't mind splitting this PR into separate commits @sebastianst, I'd be happy to give this another look.

I gave it a first pass and tried to manually diff the changes locally, but I could have missed something.

File: core/types/receipt.go (hunk 1 above)
Comment (geoknee) [reply core/types/receipt.go:617]: I think the condition is equivalent because of this: 

https://github.com/ethereum-optimism/op-geth/blob/96738d21616cd7c37101fd49234e361f70e7b64c/params/config.go#L907-L910

File: core/types/receipt_opstack.go (hunk 3 above)
Comment (geoknee) [reply core/types/receipt_opstack.go:49]: Just noting also that we should specify whatever behaviour we want to enshrine in the specs, as was done here https://specs.optimism.io/protocol/isthmus/exec-engine.html#receipts.

File: core/types/receipt_opstack.go (hunk 4 above)
Comment (geoknee) [reply core/types/receipt_opstack.go]: I think the system deposit (l1 attribs) tx would be skipped below because of the `isDepositTx` check.

I'm generally unsure about this requirement, it does match the condition in the calling function `DeriveFields()`, but I don't understand it. We already skip deposit transactions in the loop below, so I don't know why we can't just call this unconditionally (perhaps it is a performance optimisation?) And what if we had multiple deposits in a block, we would still call this function and potentially skip all the transactions in the loop.

**Comment by geoknee** [comment]
I was hoping this tool would make validating the test file migration easier https://app.semanticdiff.com/gh/ethereum-optimism/op-geth/pull/710 but it doesn't help much.

File: core/types/receipt_opstack.go (hunk 4 above)
Comment (joshklop) [reply core/types/receipt_opstack.go]: > so I don't know why we can't just call this unconditionally (perhaps it is a performance optimisation?)

It would panic since we read `txs[0].Data()`. Ofc we can guard against this by checking the length of the txs slice, but I don't want to bikeshed this comment too much: as long as it accurately documents the behavior I'm happy.

Comment (geoknee) [reply core/types/receipt_opstack.go]: Yeah completely empty tx slice would be problematic, but shouldn't happen on an optimism chain, so the tx length check in the caller seems redundant to me still. Doesn't need to block this PR since I think it preserves original semantics, but I would like to either understand this better or circle back and change it.

Comment (joshklop) [reply core/types/receipt_opstack.go]: The genesis block might be an edge case.

File: core/types/receipt.go (hunk 1 above)
Comment (sebastianst) [reply core/types/receipt.go:617]: Correct, using `config.Optimism != nil` is an outdated bad way to check for Optimism forks. In fact, we also think about removing all the `IsOptimism...` checks because any chain config with at least bedrock enabled must have a non-nil `Optimism` field.

File: core/types/receipt_opstack.go (hunk 3 above)
Comment (sebastianst) [reply core/types/receipt_opstack.go:49]: I am following the pattern to include those block level fields like we did for all previous forks. Inferring the scalar from the `BlobGasUsed` would be non-trivial, as one would have to recalculate the DA usage estimate.

I'll add this to the specs, it's indeed still missing: https://github.com/ethereum-optimism/specs/pull/839

File: core/types/receipt_opstack.go (hunk 4 above)
Comment (sebastianst) [reply core/types/receipt_opstack.go]: The edge case I wanted to cover with this condition is the Jovian activation block, because it still uses the old L1 attributes deposit. But I realize the old condition doesn't work. I improved the method to test that the last transaction is a non-deposit one at the top and exit early otherwise. This way, the function can also be called if having only a single deposit tx.

File: core/types/receipt_opstack.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+		rs[i].L1BaseFeeScalar = u32ptrTou64ptr(gasParams.l1BaseFeeScalar)
+		rs[i].L1BlobBaseFeeScalar = u32ptrTou64ptr(gasParams.l1BlobBaseFeeScalar)
+		if gasParams.operatorFeeScalar != nil && gasParams.operatorFeeConstant != nil && (*gasParams.operatorFeeScalar != 0 || *gasParams.operatorFeeConstant != 0) {
+			rs[i].OperatorFeeScalar = u32ptrTou64ptr(gasParams.operatorFeeScalar)
+			rs[i].OperatorFeeConstant = gasParams.operatorFeeConstant
+		}
+		if isJovian {
+			rs[i].DAFootprintGasScalar = &daFootprintGasScalar
+			rs[i].BlobGasUsed = daFootprintGasScalar * rcd.EstimatedDASize().Uint64()
+		}
```
[the remark this answers is not in the document]
Comment (sebastianst) [reply core/types/receipt_opstack.go:51]: dismiss
context guarantees no overflow.

**Comment by sebastianst** [comment]
> This is a bit challenging to review since there are intersecting refactors in the same commit as the consensus changes. If you don't mind splitting this PR into separate commits @sebastianst, I'd be happy to give this another look.
> 
> I gave it a first pass and tried to manually diff the changes locally, but I could have missed something.

I split up the movement of OP receipt tests to file `receipt_opstack_test.go` into a separate commit. Note that you can pretty much review in isolation the diff between `receipt_test.go` and `receipt_opstack_test.go` (first commit now, mostly moving around plus some simplifications) and then the diff between `receipt.go` and `receipt_opstack.go` (second commit, the actual changes to receipt derivation).

Note that none of this are _consensus_ changes, only changes the receipts RPC API.

File: core/types/receipt_test.go [hunk 6]
```diff
@@ -240,11 +161,13 @@ var (
...
+	blockNumber = big.NewInt(1)
+	blockTime   = uint64(2)
+	blockHash   = common.BytesToHash([]byte{0x03, 0x14})
+
 	depNonce1                   = uint64(7)
 	depNonce2                   = uint64(8)
-	blockNumber                 = big.NewInt(1)
-	blockTime                   = uint64(2)
-	blockHash                   = common.BytesToHash([]byte{0x03, 0x14})
 	canyonDepositReceiptVersion = CanyonDepositReceiptVersion
```
Comment (sebastianst) [inline core/types/receipt_test.go:171]: reduces OP diff

File: core/types/receipt_test.go [hunk 7]
```diff
@@ -493,11 +416,11 @@ func TestDecodeEmptyTypedReceipt(t *testing.T) {
...
-	baseFee := big.NewInt(1000)
+	basefee := big.NewInt(1000)
 	blobGasPrice := big.NewInt(920)
 	receipts := getTestReceipts()
 	derivedReceipts := clearComputedFieldsOnReceipts(receipts)
-	err := Receipts(derivedReceipts).DeriveFields(params.TestChainConfig, blockHash, blockNumber.Uint64(), blockTime, baseFee, blobGasPrice, txs)
+	err := Receipts(derivedReceipts).DeriveFields(params.TestChainConfig, blockHash, blockNumber.Uint64(), blockTime, basefee, blobGasPrice, txs)
 	if err != nil {
 		t.Fatalf("DeriveFields(...) = %v, want <nil>", err)
 	}
```
Comment (sebastianst) [inline core/types/receipt_test.go:426]: reduces OP diff

File: internal/ethapi/api.go [hunk 8]
```diff
@@ -1735,8 +1735,14 @@ func marshalReceipt(receipt *types.Receipt, blockHash common.Hash, blockNumber u
 		if receipt.OperatorFeeConstant != nil {
 			fields["operatorFeeConstant"] = hexutil.Uint64(*receipt.OperatorFeeConstant)
 		}
+		// Fields added in Jovian
+		if receipt.DAFootprintGasScalar != nil {
+			fields["daFootprintGasScalar"] = hexutil.Uint64(*receipt.OperatorFeeScalar)
```
Comment (joshklop) [inline internal/ethapi/api.go]: Operator fee scalar is not footprint scalar... were the ATs passing in the monorepo against this branch? If so that's a bit concerning and we should make sure nothing else is broken.

File: core/types/receipt_opstack_test.go [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+	err := Receipts(derivedReceipts).DeriveFields(bedrockGenesisTestConfig, blockHash, blockNumber.Uint64(), 0, baseFee, nil, txs)
+	require.Error(t, err)
+
+	err = Receipts(derivedReceipts).DeriveFields(ecotoneTestConfig, blockHash, blockNumber.Uint64(), 0, baseFee, nil, txs)
+	require.NoError(t, err)
+	diffReceipts(t, receipts, derivedReceipts)
+}
+
+func TestDeriveOptimismIsthmusTxReceipts(t *testing.T) {
+	// Isthmus style l1 attributes with baseFeeScalar=2, blobBaseFeeScalar=3, baseFee=1000*1e6, blobBaseFee=10*1e6, operatorFeeScalar=1439103868, operatorFeeConstant=1256417826609331460
```
Comment (joshklop) [inline core/types/receipt_opstack_test.go:266]: Why the change the in `operatorFeeScalar` and `operatorFeeConstant`?

File: core/types/receipt_opstack_test.go [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+	err := Receipts(derivedReceipts).DeriveFields(bedrockGenesisTestConfig, blockHash, blockNumber.Uint64(), 0, baseFee, nil, txs)
+	require.Error(t, err)
+
+	err = Receipts(derivedReceipts).DeriveFields(isthmusTestConfig, blockHash, blockNumber.Uint64(), 0, baseFee, nil, txs)
+	require.NoError(t, err)
+	diffReceipts(t, receipts, derivedReceipts)
+}
+
+func TestDeriveOptimismIsthmusTxReceiptsNoOperatorFee(t *testing.T) {
+	// Isthmus style l1 attributes with baseFeeScalar=2, blobBaseFeeScalar=3, baseFee=1000*1e6, blobBaseFee=10*1e6, operatorFeeScalar=0, operatorFeeConstant=0
```
Comment (joshklop) [inline core/types/receipt_opstack_test.go:288]: Nice catch in the comment here.

File: internal/ethapi/api.go (hunk 8 above)
Comment (sebastianst) [reply internal/ethapi/api.go]: Copy 🍝 ... good catch! In fact, the ATs are failing because of this.

File: core/types/receipt_opstack_test.go (hunk 9 above)
Comment (sebastianst) [reply core/types/receipt_opstack_test.go:266]: The comment was wrong, those are the actual values used. Probably some AI slop.

**Review by joshklop — APPROVED** [review]

**Review by teddyknox — APPROVED** [review]
